src/grav/test.py

```diff
@@ -1,116 +1,141 @@
-    # if the test dies partway through, the container may
-    # still be running, and should be stopped
-    inspect_path = "/containers/%s/json" % container_id
-    resp = client.get(inspect_path)
-    assert resp.status == 200
-    json = resp.json()
-    state = json["State"]
-    running = state["Running"]
-    if running:
         # stop
         stop_path = "/containers/%s/stop" % container_id
         resp = client.post(stop_path)
         resp.read()  # TODO: have this handled by the client
-        assert resp.status == 204
-
-    # delete container
-    delete_path = "/containers/" + container_id
-    resp = client.request("DELETE", delete_path)
-    resp.read()  # TODO: have this handled by the client
-    assert resp.status == 204
-
-    # delete image
-    delete_path = "/images/" + image_id
-    resp = client.request("DELETE", delete_path)
-    resp.read()  # TODO: have this handled by the client
-    assert resp.status == 200  # why isn't this 204 like the other api endpoints?
-
+        self.assertStatus(resp, 204)
+
+        # logs
+        logs_path = "/containers/%s/logs" % container_id
+        logs_path += "?stdout=true&stderr=true"
+        resp = client.get(logs_path)
+        self.assertStatus(resp, 200)
+        # logs doesn't return json, but a multiplexed stream
+        stdout, stderr = demux_logs(resp.read())
+
+        # validate
+        stdout_lines = stdout.splitlines()
+        self.assertEqual(stdout_lines[0], "Hello Gravitational!")
+        self.assertEqual(stdout_lines[-1], "Terminated. Bye!")
+        self.assertEqual(stderr, "")
+
+    def tearDown(self):
+        client = self.client
+        container_id = self.container_id
+        image_id = self.image_id
+
+        # if the test dies partway through, the container may
+        # still be running, and should be stopped
+        inspect_path = "/containers/%s/json" % container_id
+        resp = client.get(inspect_path)
+        assert_status(resp, 200)
+        json = resp.json()
+        state = json["State"]
+        running = state["Running"]
+        if running:
+            # stop
+            stop_path = "/containers/%s/stop" % container_id
+            resp = client.post(stop_path)
+            resp.read()  # TODO: have this handled by the client
+            assert_status(resp, 204)
+
+        # delete container
+        delete_path = "/containers/" + container_id
+        resp = client.request("DELETE", delete_path)
+        resp.read()  # TODO: have this handled by the client
+        assert_status(resp, 204)
 
-if __name__ == "__main__":
-    client, image_id, container_id = setup()
-    test(client, container_id)
-    teardown(client, container_id, image_id)
+        # delete image
+        delete_path = "/images/" + image_id
+        resp = client.request("DELETE", delete_path)
+        resp.read()  # TODO: have this handled by the client
+        # why isn't this 204 like the other api endpoints without a body?
```

Comment:
https://docs.docker.com/engine/api/v1.40/#operation/ImageDelete
Looks like DELETE /images/{id} returns an array of deleted image ids

Reply:
This is a great idea, but I'm going to leave it for the time being because it isn't as high priority to this project as getting the readme & last of the build work done.

Here is why: a typical response json looks like:

```
[
  {'Deleted': 'sha256:3ca1723f55b595b5646967b8b333be735cda2846c77b0f73fd606d8fb6f55f16'},
  {'Deleted': 'sha256:b208507827cb7c5b61fda4b1901a12a7d7d62ef5c81105af32a42de96d247b55'},
  {'Deleted': 'sha256:03901b4a2ea88eeaad62dbe59b072b28b6efa00491962b8741081c5df50c65e0'
]
```

There may be `Untagged`keys in those single item maps too, as well as other stuff?  I'd need to spend some more time reading the docs.  Given this format there isn't a concise known path I could pull out of the response (`json[0]['Deleted']` is a guess, but I wouldn't bank on the determinism of that). To do it right I would instead need some minor scanning logic to verify image deletion.

Reply:
#10 filed for this to be dealt with in future work.